### PR TITLE
*: fix locking operation and add tests

### PR DESCRIPTION
**Pull request by jonboulle**
Deprecates #16.

### CODE REVIEW

**Comment by jonboulle** [comment]
GitHub is not showing these in the correct order. I rebased to order the tests before Dan's fix.

File: iptables/iptables_test.go [hunk 1]
```diff
@@ -30,126 +31,157 @@ func randChain(t *testing.T) string {
...
-	err = ipt.ClearChain("filter", chain)
-	if err != nil {
-		t.Fatalf("Failed to clear test chain: %v", err)
-	}
-
-	// Delete the chain that was created
-	err = ipt.DeleteChain("filter", chain)
-	if err != nil {
-		t.Fatalf("Failed to delete test chain: %v", err)
+	for _, ipt := range mustTestableIptables() {
```
Comment (steveeJ) [inline iptables/iptables_test.go]: Could you make a function out of the loop body and pass `ipt` to it? that way we also don't have so much code difference that is mainly due to indentation.

File: iptables/iptables_test.go [hunk 2]
```diff
@@ -30,126 +31,157 @@ func randChain(t *testing.T) string {
...
+	}
+	return ipts
+}
 
-	// chain empty, should be ok
-	err = ipt.DeleteChain("filter", newChain)
-	if err != nil {
-		t.Fatalf("DeleteChain of empty chain failed: %v", err)
+func TestChain(t *testing.T) {
+	for _, ipt := range mustTestableIptables() {
```
Comment (steveeJ) [inline iptables/iptables_test.go]: see below

File: iptables/iptables.go [hunk 3]
```diff
@@ -81,10 +73,11 @@ func (ipt *IPTables) Exists(table, chain string, rulespec ...string) (bool, erro
 	}
 	cmd := append([]string{"-t", table, "-C", chain}, rulespec...)
 	err := ipt.run(cmd...)
```
Comment (steveeJ) [inline iptables/iptables.go]: can you do

```
err, eok := ipt.run(cmd...).(*Error)
```

?

Comment (jonboulle) [inline iptables/iptables.go]: No.
https://github.com/coreos/go-iptables/pull/19/files/6bc61e78c75ea00e1ef1638d8dc7e192b8d914bf..696a9460abf9e82549574025e027c01e5abe9964#diff-b90cadcd0928c1e490272f4761a52bacL90

**Comment by jonboulle** [comment]
@steveeJ ok for this to land?

File: iptables/iptables_test.go [hunk 4]
```diff
@@ -30,16 +31,46 @@ func randChain(t *testing.T) string {
...
+	// ensure we check one variant without built-in locking
+	if ipt.hasWait {
+		iptNoWait := &IPTables{
+			path:    ipt.path,
+			hasWait: false,
+		}
+		ipts = append(ipts, iptNoWait)
+	}
+	// ensure we check one variant without built-in checking
+	if ipt.hasCheck {
```
Comment (jonboulle) [inline iptables/iptables_test.go]: If iptables doesn't have a feature we can't check that feature..

**Pull request merged**